precompiles/zk-groth16-verify/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use pallet_evm::{ExitError, ExitSucceed, LinearCostPrecompile, PrecompileFailure};
+use sp_core::U256;
+
+use sp_std::vec::Vec;
+
+use ark_crypto_primitives::snark::SNARK;
+use ark_groth16::Groth16;
+
+use ark::{ark_bn254_fr, ark_bn254_g1, ark_bn254_g2};
+use precompile_utils::EvmDataWriter;
+
+pub struct ZKGroth16Verify;
+
+mod ark;
+
+impl LinearCostPrecompile for ZKGroth16Verify {
+	const BASE: u64 = 60;
+	const WORD: u64 = 12;
+
+	fn execute(input: &[u8], _cost: u64) -> Result<(ExitSucceed, Vec<u8>), PrecompileFailure> {
+		log::trace!(target: "precompiles::zk_groth16_verify::execute", "In zk-groth16-verify");
+		const MIN_INPUT_LENGTH: usize = 24 * 32;
+		if input.len() < 4 + MIN_INPUT_LENGTH {
+			return Err(PrecompileFailure::from(ExitError::InvalidRange));
+		}
+		let (_selector, input_stripped) = input.split_at(4);
+
+		let mut cursor = 0;
+		let mut next = || {
+			let start = cursor;
+			cursor += 32;
+			&input_stripped[start..cursor]
+		};
+
+		let proof_a = ark_bn254_g1(next(), next())
+			.ok_or_else(|| PrecompileFailure::from(ExitError::InvalidRange))?;
+		let proof_b = ark_bn254_g2(next(), next(), next(), next())
+			.ok_or_else(|| PrecompileFailure::from(ExitError::InvalidRange))?;
+		let proof_c = ark_bn254_g1(next(), next())
+			.ok_or_else(|| PrecompileFailure::from(ExitError::InvalidRange))?;
+
+		let proof: ark_groth16::Proof<ark_ec::bn::Bn<ark_bn254::Config>> = ark_groth16::Proof {
+			a: proof_a,
+			b: proof_b,
+			c: proof_c,
+		};
+
```

Comment:
I would add intermediate logs. Like parsed proof, parsed key to localize failure point in debugging

Reply:
Resolved.